clippy_utils/src/hir_utils.rs

```diff
@@ -288,21 +290,54 @@ impl HirEqInterExpr<'_, '_, '_> {
                     && self.eq_expr(l_receiver, r_receiver)
                     && self.eq_exprs(l_args, r_args)
             },
+            (ExprKind::Path(l), ExprKind::Path(r)) => self.eq_qpath(l, r),
             (&ExprKind::Repeat(le, ll), &ExprKind::Repeat(re, rl)) => {
                 self.eq_expr(le, re) && self.eq_array_length(ll, rl)
             },
             (ExprKind::Ret(l), ExprKind::Ret(r)) => both(l, r, |l, r| self.eq_expr(l, r)),
-            (ExprKind::Path(l), ExprKind::Path(r)) => self.eq_qpath(l, r),
             (&ExprKind::Struct(l_path, lf, ref lo), &ExprKind::Struct(r_path, rf, ref ro)) => {
                 self.eq_qpath(l_path, r_path)
                     && both(lo, ro, |l, r| self.eq_expr(l, r))
                     && over(lf, rf, |l, r| self.eq_expr_field(l, r))
             },
             (&ExprKind::Tup(l_tup), &ExprKind::Tup(r_tup)) => self.eq_exprs(l_tup, r_tup),
             (&ExprKind::Unary(l_op, le), &ExprKind::Unary(r_op, re)) => l_op == r_op && self.eq_expr(le, re),
-            (&ExprKind::Array(l), &ExprKind::Array(r)) => self.eq_exprs(l, r),
-            (&ExprKind::DropTemps(le), &ExprKind::DropTemps(re)) => self.eq_expr(le, re),
-            _ => false,
+            // Else branches for branches above, grouped as per `match_same_arms`.
+            (&ExprKind::AddrOf(..)
+            | &ExprKind::Array(..)
+            | &ExprKind::Assign(..)
+            | &ExprKind::AssignOp(..)
+            | &ExprKind::Binary(..)
+            | &ExprKind::Block(..)
+            | &ExprKind::Box(..)
+            | &ExprKind::Break(..)
+            | &ExprKind::Call(..)
+            | &ExprKind::Cast(..)
+            | &ExprKind::Continue(..)
+            | &ExprKind::DropTemps(..)
+            | &ExprKind::Field(..)
+            | &ExprKind::Index(..)
+            | &ExprKind::If(..)
+            | &ExprKind::Let(..)
+            | ExprKind::Lit(..)
+            | &ExprKind::Loop(..)
+            | &ExprKind::Match(..)
+            | &ExprKind::MethodCall(..)
+            | ExprKind::Path(..)
+            | &ExprKind::Repeat(..)
+            | ExprKind::Ret(..)
+            | &ExprKind::Struct(..)
+            | &ExprKind::Tup(..)
+            | &ExprKind::Unary(..)
+
+            // Note: The following do NOT have a positive match.
+            | &ExprKind::ConstBlock(_)
+            | &ExprKind::Closure(_)
+            | &ExprKind::InlineAsm(_)
+            | &ExprKind::Yield(..)
+            | &ExprKind::Type(..)
```

Comment:
Those are the reason, for which I opened the issue.  Can you try to add an implementation for those? If you're unsure about something just ask and we can come up with something together.

Reply:
I'm really not sure what definition of equality we're implementing. Can two closures even be equal? Or only if they have the same `def_id`?

Reply:
An "implementation" for those that is just `(Closure, Closure) => false,` is fine. But I think 2 closures can be equal. E.g.:

```rust
let x = || 4 + 2;
let y = || 4 + 2;
```

The closures have the exact same expressions inside and the same argument count and everything. Only their span is different. Otherwise you could just remove `y` and use `x` everywhere instead.

Reply:
So we pretty much care about the literal content of the expressions if they are equal. (Disregarding formatting and spans)

Reply:
I would just do this in one PR if you don't mind.

Reply:
> An "implementation" for those that is just `(Closure, Closure) => false,` is fine. But I think 2 closures can be equal. E.g.:
> 
> ```rust
> let x = || 4 + 2;
> let y = || 4 + 2;
> ```
> 
> The closures have the exact same expressions inside and the same argument count and everything. Only their span is different. Otherwise you could just remove `y` and use `x` everywhere instead.

Not sure I follow. Does this mean that closure comparison use De Bruijn indices? Feels a bit overkill for what looked like a style PR :)